core/src/main/java/io/questdb/log/LogAlertSocketWriter.java

```diff
@@ -0,0 +1,266 @@
+        socket.connect();
+    }
+
+    @Override
+    public void close() {
+        if (socket != null) {
+            socket.close();
+        }
+    }
+
+    @Override
+    public boolean runSerially() {
+        return writeSequence.consumeAll(alertsSource, alertsProcessor);
+    }
+
+    @VisibleForTesting
+    void setBufferSize(String bufferSize) {
+        this.bufferSize = bufferSize;
+    }
+
+    @VisibleForTesting
+    void setLocation(String location) {
+        this.location = location;
+    }
+
+    @VisibleForTesting
+    String getLocation() {
+        return location;
+    }
+
+    @VisibleForTesting
+    void setSocketAddress(String socketAddress) {
+        this.socketAddress = socketAddress;
+    }
+
+    @VisibleForTesting
+    String getSocketAddress() {
+        return socket.getSocketAddress();
+    }
+
+    @VisibleForTesting
+    HttpLogAlertBuilder getAlertBuilder() {
+        return alertBuilder;
+    }
+
+    @VisibleForTesting
+    int getBufferSize() {
+        return socket.getOutBufferSize();
+    }
+
+    private void loadLogAlertTemplate() {
+        final long now = clock.getTicks();
+        if (location.isEmpty()) {
+            location = DEFAULT_ALERT_TPT_FILE;
+        }
+        location = dollar$.resolveEnv(location, now).toString();
+
+        // read template, resolve env vars within (except $ALERT_MESSAGE)
+        boolean wasRead = false;
+        try (InputStream is = LogAlertSocketWriter.class.getResourceAsStream(location)) {
+            if (is != null) {
+                byte[] buff = new byte[LogAlertSocket.IN_BUFFER_SIZE];
+                int n = is.read(buff, 0, buff.length);
+                if (n > 0) {
+                    dollar$.resolve(new String(buff, 0, n, Files.UTF_8), now, alertProps);
+                    wasRead = true;
+                }
+            }
+        } catch (IOException e) {
+            // it was not a resource ("/resource_name")
+        }
+        if (!wasRead) {
+            dollar$.resolve(
+                    readFile(
+                            location,
+                            socket.getInBufferPtr(),
+                            socket.getInBufferSize(),
+                            ff
+                    ),
+                    now,
+                    alertProps
+            );
+        }
+
+        // consolidate/check/load template to the outbound socket buffer
+        dollar$.resolve(dollar$.toString(), now, alertProps);
+        ObjList<Sinkable> components = dollar$.getLocationComponents();
+        if (dollar$.getKeyOffset(MESSAGE_ENV) < 0 || components.size() < 3) {
+            throw new LogError(String.format(
+                    "Bad template, no %s declaration found %s",
+                    MESSAGE_ENV_VALUE,
+                    location));
+        }
+        alertBuilder = new HttpLogAlertBuilder(socket)
+                .putHeader(LogAlertSocket.localHostIp)
+                .put(components.getQuick(0))
```

Comment:
it is not clear to me why static part of components is only one item 0 ?

```
hello ${x} i am ${y} here is $(message), cool?
```

components 0-4 are static, 5 is message, 6 is footer

Reply:
two passes on processing the template, on first pass we resolve all dollar expressions and leave ${ALERT_MESSAGE} intact. Then we call toString() and feed this again to dollar$ to ensure there are only two static blocks and one variable block (${ALERT_MESSAGE}).